org.eclipse.lsp4j.jsonrpc/src/main/java/org/eclipse/lsp4j/jsonrpc/json/MessageJsonHandler.java

```diff
@@ -74,19 +103,39 @@ public void setMethodProvider(MethodProvider methodProvider) {
 		this.methodProvider = methodProvider;
 	}
 	
-	public Message parseMessage(CharSequence input) {
-		return gson.fromJson(input.toString(), Message.class);
+	public Message parseMessage(CharSequence input) throws JsonParseException {
+		StringReader reader = new StringReader(input.toString());
+		return parseMessage(reader);
 	}
 	
-	public Message parseMessage(Reader input) {
-		return gson.fromJson(input, Message.class);
+	public Message parseMessage(Reader input) throws JsonParseException {
+		JsonReader jsonReader = new JsonReader(input);
+		Message message = gson.fromJson(jsonReader, Message.class);
```

Comment:
Seems odd to catch only the case where Input isn’t fully consumed, yet an invalid JSON string isn’t caught at all.

Reply:
An invalid JSON string is caught in MessageTypeAdapter. We can't catch it here because we need an at least partially constructed Message instance in order to be able to respond to it.